### PR TITLE
BUG: _validate_setitem_value fails to raise for PandasArray #51044

### LINKED ISSUE #51044: BUG: _validate_setitem_value fails to raise for PandasArray
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
arr = pd.Series(range(5)).array

arr._validate_setitem_value("foo")  # <- fails to raise
arr._validate_setitem_value(1.5)  # <- fails to raise
```

### Issue Description

AFAICT this doesn't affect anything bc we always extract PandasArray before doing anything.

Note that if this _did_ raise on 1.5, it might cause a problem with `arr.searchsorted(floats)` raising when we wouldn't want it to.

### Expected Behavior

NA

### Installed Versions

<details>

Replace this line with the output of pd.show_versions()

</details>

### PR DESCRIPTION
**Pull request by ArchPh03nix**
- [x] closes #51044 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by ArchPh03nix** [comment]
I would like to know if I'm going in the right direction/if the logic needs to be changed

File: pandas/core/arrays/numpy_.py [hunk 1]
```diff
@@ -507,6 +507,23 @@ def to_numpy(
...
+        kind = self.dtype.kind
+        if kind == "b":
+            if lib.is_bool(value):
+                return value
+
+        elif kind == "f":
+            if lib.is_integer(value) or lib.is_float(value):
+                return value
+        else:
+            if lib.is_integer(value) or (lib.is_float(value) and value.is_integer()):
```
Comment (jbrockmendel) [inline pandas/core/arrays/numpy_.py:520]: this is only for integer-dtype.  also you'll need to check itemsize to be sure it fits.  there is probably something in core.dtypes.cast (maybe np_can_hold_element?) that can be reused here

File: pandas/core/arrays/numpy_.py [hunk 2]
```diff
@@ -507,6 +507,23 @@ def to_numpy(
...
 
+    def _validate_setitem_value(self, value):
+        kind = self.dtype.kind
+        if kind == "b":
+            if lib.is_bool(value):
+                return value
+
+        elif kind == "f":
+            if lib.is_integer(value) or lib.is_float(value):
+                return value
```
Comment (jbrockmendel) [inline pandas/core/arrays/numpy_.py:518]: these checks are correct for scalar value, not sure if validate_setitem_value needs to handle listlike?

**Comment by jbrockmendel** [comment]
> I would like to know if I'm going in the right direction/if the logic needs to be changed

A bunch of tests are failing on the CI.  This is the first place to look for potential problems.

File: pandas/core/arrays/numpy_.py [hunk 3]
```diff
@@ -507,6 +507,46 @@ def to_numpy(
 
         return result
 
+    def _validate_setitem_value(self, value):
+        if type(value) == int:
+            if (
```
Comment (dbalagula) [inline pandas/core/arrays/numpy_.py:512]:
```
if self.dtype in [NumpyEADtype("int64"), NumpyEADtype("float64"), NumpyEADtype("uint16"), NumpyEADtype("object")]
```

you can do the same for the other branches as well

File: pandas/core/arrays/numpy_.py [hunk 4]
```diff
@@ -507,6 +507,46 @@ def to_numpy(
...
+                return value
+        elif type(value) == str:
+            if (
+                self.dtype == NumpyEADtype("str")
+                or self.dtype == NumpyEADtype("object")
+                or self.dtype is None
+                or self.dtype == NumpyEADtype("U32")
+            ):
+                return value
+        elif NumpyEADtype(type(value)) == NumpyEADtype(self.dtype) or NumpyEADtype(
```
Comment (dbalagula) [inline pandas/core/arrays/numpy_.py:535]: It looks like you have a check for

```
NumpyEADtype(type(value)) == NumpyEADtype(type(value))
```

you can just change it to 

```
elif NumpyEADtype(type(value)) == NumpyEADtype(self.dtype):
```

File: pandas/core/arrays/numpy_.py [hunk 5]
```diff
@@ -507,6 +507,46 @@ def to_numpy(
...
+            type(value)
+        ) == NumpyEADtype(type(value)):
+            return value
+        else:
+            if self.dtype == NumpyEADtype("object") or self.dtype is None:
```
Comment (dbalagula) [inline pandas/core/arrays/numpy_.py:540]: you can simplify to 

```
elif self.dtype == NumpyEADtype("object") or self.dtype is None
```

File: pandas/core/arrays/numpy_.py [hunk 6]
```diff
@@ -507,6 +507,46 @@ def to_numpy(
...
+                return value
+        if NumpyEADtype(type(value)) != self.dtype:
```
Comment (dbalagula) [inline pandas/core/arrays/numpy_.py:542]: I think you can just add a check for `NumpyEADtype(type(value)) == self.dtype` as an `or` when you do `NumpyEADtype(type(value)) == NumpyEADtype(self.dtype)`

Comment (dbalagula) [reply pandas/core/arrays/numpy_.py:542]: Then you can just raise a TypeError without any more checks

File: pandas/core/arrays/numpy_.py (hunk 2 above)
Comment (ArchPh03nix) [reply pandas/core/arrays/numpy_.py:534]: The TypeError still occurs if listlike values are entered in to validate causing tests to fail, even if it's supposed to work, see pandas/tests/arrays/numpy_/test_numpy.py test_ufunc_unary

**Comment by dbalagula** [comment]
Hi @mroeschke are you able to provide any guidance here?

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Pull request closed without merging**